### PR TITLE
Fix Viewer API definitions and include in CI

### PR DESCRIPTION
**Pull request by michael-yx-wu**
The Viewer API definitions do not compile because of missing imports and
anonymous objects are typed as `Object`. These issues were not caught
during CI because the test project was not compiling anything from the
Viewer API.

As an example of the first problem:

```
/**
 * @implements MyInterface
 */
export class MyClass {
    ...
}
```

will generate a broken definition that doesn’t import MyInterface:

```
/**
 * @implements MyInterface
 */
export class MyClass implements MyInterface {
    ...
}
```

This can be fixed by adding a typedef jsdoc to specify the import:

```
/** @typedef {import("./otherFile").MyInterface} MyInterface */
```

See https://github.com/jsdoc/jsdoc/issues/1537 and
https://github.com/microsoft/TypeScript/issues/22160 for more details.

As an example of the second problem:

```
/**
 * Gets the size of the specified page, converted from PDF units to inches.
 * @param {Object} An Object containing the properties: {Array} `view`,
 *   {number} `userUnit`, and {number} `rotate`.
 */
function getPageSizeInches({ view, userUnit, rotate }) {
    ...
}
```

generates the broken definition:

```
function getPageSizeInches({ view, userUnit, rotate }: Object) {
    ...
}
```

The jsdoc should specify the type of each nested property:

```
/**
 * Gets the size of the specified page, converted from PDF units to inches.
 * @param {Object} options An object containing the properties: {Array} `view`,
 *   {number} `userUnit`, and {number} `rotate`.
 * @param {number[]} options.view
 * @param {number} options.userUnit
 * @param {number} options.rotate
 */
```

### CODE REVIEW

File: web/annotation_layer_builder.js
Comment (Snuffleupagus) [inline web/annotation_layer_builder.js:17]: Here, and elsewhere, you really ought to put a blank line between this sort of comment and the imports below since as-is this is *somewhat* hurting readability.

File: web/interfaces.js
Comment (Snuffleupagus) [inline web/interfaces.js:111]: Why `*` here, since it should be a string!?

File: web/interfaces.js
Comment (Snuffleupagus) [inline web/interfaces.js:113]: Also, both of these should be marked as optional (since they are).

Please see https://github.com/mozilla/pdf.js/blob/bf5a45ce6d0e9d070204859a39844bcdaf69e9aa/web/pdf_history.js#L36-L42 for the expected types (and what's optional).

File: web/interfaces.js
Comment (Snuffleupagus) [inline web/interfaces.js:123]: Again, please see https://github.com/mozilla/pdf.js/blob/bf5a45ce6d0e9d070204859a39844bcdaf69e9aa/web/pdf_history.js#L44-L50 for the expected types.

File: web/pdf_rendering_queue.js
Comment (Snuffleupagus) [inline web/pdf_rendering_queue.js:38]: Sorry, but why?

This property really *should* be completely internal.

File: web/ui_utils.js
Comment (Snuffleupagus) [inline web/ui_utils.js:291]: Please add a blank line here.

File: web/ui_utils.js
Comment (Snuffleupagus) [inline web/ui_utils.js:290]:
```javascript
  * @property {number} width - In inches.
  * @property {number} height - In inches.
```

File: web/ui_utils.js
Comment (Snuffleupagus) [inline web/ui_utils.js:300]: Let's introduce a proper `typedef` instead here, i.e. something like:
```javascript
/**
 * @typedef {Object} getPageSizeInchesParameters
 * @property {Array} view
 * @property {number} userUnit
 * @property {number} rotation
 */

/**
 * Gets the size of the specified page, converted from PDF units to inches.
 * 
 * @param {getPageSizeInchesParameters} params
 * @returns {PageSize}
 */
```

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
I don't really understand the "add CI check task"-part of the commit message, is there something missing in the patch?
Also, it probably wouldn't hurt to actually mention (in the first line of the commit message) that it's the TypeScript definitions for the *viewer* that's being changed here.

File: web/interfaces.js
Comment (michael-yx-wu) [reply web/interfaces.js:111]: Ah thanks, I wasn't sure about the type so I left it as `*`. Will change!

File: web/interfaces.js
Comment (Snuffleupagus) [inline web/interfaces.js:161]: This also looks somewhat weird, since when `resume` is defined it's a regular function and *not* a getter.

Comment (michael-yx-wu) [reply web/interfaces.js:161]: Agree, but in the implementing classes (e.g. `PDFPageView`) , it's actually initialized to be `null`

https://github.com/mozilla/pdf.js/blob/bf5a45ce6d0e9d070204859a39844bcdaf69e9aa/web/pdf_page_view.js#L116

and later set to a function

https://github.com/mozilla/pdf.js/blob/bf5a45ce6d0e9d070204859a39844bcdaf69e9aa/web/pdf_page_view.js#L571

which causes the generated types for `PDFPageView` and this interface to disagree unless I type `resume` like this. I felt like it'd be safer for this PR to have the types reflect the current implementation rather than change the implementation to match the desired types.

Comment (Snuffleupagus) [reply web/interfaces.js:161]: > which causes the generated types for `PDFPageView` and this interface to disagree unless I type `resume` like this.

But this actually means the interface-definition is now by definition *incorrect*, which really doesn't help someone trying to see how it should be correctly implemented. Can't we just tag the method as follows instead?
```javascript
/**
 * @ignore
 */
resume() {}
```

File: web/pdf_rendering_queue.js
Comment (michael-yx-wu) [reply web/pdf_rendering_queue.js:38]: because of the `setTimeout` call, `tsc` infers this type to be `idleTimeout: NodeJS.Timeout | null;`, resulting in the following compilation error since the types test project isn't building for node:
```
Couldn't compile TypeScript test: build/typestest/types/web/pdf_rendering_queue.d.ts(9,18): error TS2503: Cannot find namespace 'NodeJS'.
```

https://github.com/mozilla/pdf.js/blob/56e7bb626c89282f207b585ebed74c767c68db95/web/pdf_rendering_queue.js#L95-L97

Writing this has made me realize that pdf.js seems to support running server-side and since `NodeJS.setTimeout` returns an object while` window.setTimeout` returns a number, it would be more proper to type this as `ReturnType<typeof setTimeout>`, where `setTimeout`'s type will ultimately depend on your project setup.

File: web/interfaces.js
Comment (michael-yx-wu) [reply web/interfaces.js:161]: `tsc` seems to be ignoring the `@ignore` annotation and generates types anyway, so I either have to set the type to describe the current situation or fix the underlying implementations to match the desired type of `() => void`. I'll make an attempt at a fix since it looks relatively simple

File: test/types/main.ts
Comment (michael-yx-wu) [inline test/types/main.ts:2]: The contents of the `typestest` folder is somewhat different from `pdfjs-dist`. Notably, it's missing a `legacy` folder so it's not possible to check that the `external/dist/legacy/web/pdf_viewer.d.ts` file is doing the right thing during CI. Could be good to do in a future change!

Comment (Snuffleupagus) [reply test/types/main.ts:2]: The TypeScript definitions provided can be seen as a best-effort solution, hence we don't want to *further* increase test run-time for `typestest` by testing e.g. the `legacy`-folder (those builds are only provided for older environments anyway).

File: web/pdf_page_view.js
Comment (Snuffleupagus) [inline web/pdf_page_view.js:117]: Sorry, but please revert all changes of this kind throughout the patch!

We really don't want to add a, possibly small but nonetheless, overhead to our JavaScript code just for the TypeScript definitions.
Having the TypeScript definitions dictate how we can/cannot write JavaScript code isn't really OK as far as I'm concerned.

File: web/pdf_history.js
Comment (Snuffleupagus) [inline web/pdf_history.js:36]: Please revert all changes in this file, since it really doesn't seem great having to look in a different file to find the docs.

File: web/pdf_rendering_queue.js
Comment (Snuffleupagus) [inline web/pdf_rendering_queue.js:37]: Note that the code in the `web/`-folder is only really intended to run in browsers, and please make this one-line comment instead; i.e. something like `/** @type ... */`

File: web/text_layer_builder.js
Comment (Snuffleupagus) [inline web/text_layer_builder.js:17]: Missing a blank line here.

File: web/ui_utils.js
Comment (Snuffleupagus) [inline web/ui_utils.js:296]: `* @property {number} width - In inches.` and similar below.

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]

File: web/pdf_history.js
Comment (michael-yx-wu) [reply web/pdf_history.js:36]: I was worried about the duplicate definitions diverging over time. Happy to revert though

File: web/pdf_page_view.js
Comment (michael-yx-wu) [reply web/pdf_page_view.js:117]: Sounds good. But this means I'll have to also revert IRenderableView to the below to get the typings to compile:

```
class IRenderableView {
  ...
  /**
   * @type {?(() => void)}
   */
  get resume() {}
}
``` 

Is that okay?

File: web/pdf_history.js
Comment (michael-yx-wu) [reply web/pdf_history.js:36]: I suppose the types CI will catch incompatible definitions

File: web/pdf_page_view.js
Comment (Snuffleupagus) [reply web/pdf_page_view.js:117]: No, as was mentioned previously that's also not correct since this isn't a getter.

Comment (michael-yx-wu) [reply web/pdf_page_view.js:117]: What about something like this? 
```
class IRenderableView {
  constructor() {
    /** @type {(() => void) | null} */
    this.resume = null;
  }
  ...
}
```

File: web/pdf_rendering_queue.js
Comment (michael-yx-wu) [reply web/pdf_rendering_queue.js:37]: oh, in that case I will replace the `setTimeout` with `window.setTimeout` to disambiguate that we want the browser typings

File: web/pdf_rendering_queue.js
Comment (Snuffleupagus) [inline web/pdf_rendering_queue.js:99]: Please revert this change, since it'd be a case of the TypeScript definitions dictating how we can/can't write JavaScript code. Adding a one-liner comment with correct `type`-information actually seem preferable here.

Note also that we've got `setTimeout`-calls in various parts of the code-base and having one special case like this doesn't feel right (and I *really* don't want to convert all of them).

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]

**Review by timvandermeij — COMMENTED** [review]
Looks good with these comments addressed (and some of the remaining ones above). Thank you for improving the types and their testing!

**Review by Snuffleupagus — APPROVED** [review]
This looks much better, thank you!
(I'll let Tim take one last look as well, before we merge this.)

**Review by timvandermeij — APPROVED** [review]

**Pull request merged**
